workflow/controller/operator.go

```diff
@@ -1549,7 +1550,7 @@ func (woc *wfOperationCtx) executeTemplate(ctx context.Context, nodeName string,
 		memoizationStatus := &wfv1.MemoizationStatus{
 			Hit:       hit,
 			Key:       processedTmpl.Memoize.Key,
```

Comment:
Possible oversight here: if we use workflow-level memoization (and therefore no individual keys are defined), this would currently return a nil dereference or at empty string.

To solve this we would have to generate a template key name for the user. We would need to find a good way to do this, because cache keys usually come with "versions" (e.g. `cache-foo-bar-v1`) and when users want to invalidate/use new caches they just change `v1` to `v2`. If we fill in their keys for them they lose this control.

When we first designed the Memoization feature, we considered just using auto generated keys, but for this reason we decided against it.

I don't think we would want auto generated keys in this case either, so we would need to come up with a way for the user to still have control over their keys when using this f.eature

Reply:
We will only get here when `if node == nil && processedTmpl.Memoize != nil && processedTmpl.Memoize.Key != ""` met so the key must exist for a template to use memoization, either at template-level or workflow-level. In other words, users have to specify key manually in order to use memoize.

Reply:
Ah, thanks for the clarification. So to clarify: all this PR does is save the user form having to specify `cache` and `maxAge` from each template, but they still need to specify `memoization`? If that's the case I'm not sure I would think this change would be worth the new type intrusions.

Reply:
Yes but there's usability benefits. The workflow spec can also get extremely large when memoize is specified at 80% of the templates for a workflow with large number of steps.

Reply:
I just read through this again. I agree with you that this PR as it stands may not bring much benefit. I am thinking of making the following changes:
1. If users wants to use workflow-level memoization, they can set cache/maxAge/key at workflow level that will be used by all templates. The `key` they specify at workflow-level will be used as a suffix so that each template's cache key will be `templateName-key`(e.g. `cache-foo-bar-v1` where `v1` is the workflow-level key).
2. Users can override template-level cache keys.
3. Users can toggle whether to/not to use memoization at a particular template, e.g. when `key=""`. 

WDYT?

Reply:
I'll bring it up with the team

Reply:
@terrytangyuan the team discussed this and we agreed to sit on this for a while. It's low priority and we might choose to solve this another way

Reply:
No problem. Closing this for now. Please keep me posted.